lookout/style/format/feature_extractor.py

```diff
@@ -77,58 +92,84 @@ def __init__(self, *, language: str, left_siblings_window: int, right_siblings_w
-
-        self._feature_layout = OrderedDict()
-        populate_layout(FeatureGroup.node, self.node_features, None)
-        populate_layout(FeatureGroup.left, self.left_features, self.left_siblings_window)
-        populate_layout(FeatureGroup.right, self.right_features, self.right_siblings_window)
-        populate_layout(FeatureGroup.parents, self.parent_features, self.parents_depth)
-
-        self._feature_names = tuple(name for names in self.feature_layout.values()
-                                    for name in names)
-        self._selected_feature_names = (
-            self.feature_names if self.selected_features is None
-            else [self._feature_names[x] for x in self.selected_features])
-        self._feature2index = {name: i for i, name in enumerate(self.feature_names)}
-
-    @property
-    def feature_layout(self):
-        return self._feature_layout
+        if self.selected_features is not None:
+            selected_features_set = set(self.selected_features)
+
+        def populate_indices(feature_group: FeatureGroup, feature_names: Sequence[str],
+                             repetitions: int, offset: int) -> int:
+            self._feature_to_index[feature_group] = []
+            self._feature_offsets[feature_group] = []
+            for repetition in range(repetitions):
+                self._feature_to_index[feature_group].append(OrderedDict())
+                self._feature_offsets[feature_group].append(OrderedDict())
+                for feature_name in feature_names:
+                    self._feature_offsets[feature_group][repetition][feature_name] = offset
+                    feature = self._features[feature_name]
+                    names = list(feature.names)
+                    self._feature_to_index[feature_group][repetition][feature_name] = []
+                    for index in range(len(list(self._features[feature_name].names))):
+                        if (self.selected_features is None
+                                or offset in selected_features_set):
+                            if isinstance(feature, MultipleValuesFeature):
+                                name = "%s, %s" % (feature.id.name, names[index])
+                            else:
+                                name = feature.id.name
+                            self._feature_names.append("(%s, %d, %s)" % (feature_group.name,
+                                                                         repetition, name))
+                            self._feature_to_index[feature_group][repetition][feature_name].append(
+                                offset)
+                            self._index_to_feature.append((feature_group, repetition, feature_name,
+                                                           index))
+                        offset += 1
+            return offset
+        self._index_to_feature = []  # type: IndexToFeature
+        self._feature_to_index = OrderedDict()  # type: FeatureToIndex
```

Comment:
Why do we need ordered dicts?

Reply:
Goal was to be able to reproduce the feature layout from any of those but indeed it didn't prove necessary in current code. since the overhead should be really small I'd still rather keep it at is to make the codebase less error-prone. WDYT @vmarkovtsev?

Reply:
OK. Please write this in a comment above the line :)